src/phylo_flags.hpp

```diff
@@ -0,0 +1,254 @@
+  static inline PhyloFlagName run_defaults_ = {"RUN_DEFAULTS", "run_defaults", false,
+                                               PhyloFlagName::FlagType::RunAll};
+  // List of all possible options user can set.
+  std::set<PhyloFlagName> all_options_ = {run_defaults_};
+  // Flag to String.
+  std::string ToString();
+};
+
+// ** Function options for FatBeagle::Gradient().
+class PhyloFlagGradientOptions : public PhyloFlagOptions {
+ public:
+  static PhyloFlagGradientOptions CreatePhyloFlagGradientOptions();
+
+ public:
+  // Include options.
+  PhyloFlagName clock_model_rates_;
+  PhyloFlagName ratios_root_height_;
+  PhyloFlagName site_model_parameters_;
+  PhyloFlagName substitution_model_rates_;
+  PhyloFlagName substitution_model_frequencies_;
+  PhyloFlagName substitution_model_;
+  // Exclude options.
+  PhyloFlagName exclude_log_det_jacobian_gradient_;
+  // SetValue options.
+  PhyloFlagName set_gradient_delta_;
+  // Set of All Gradient options.
+  std::set<PhyloFlagName> all_options_;
+};
+
+// ** Function options for FatBeagle::LogLikelihood().
+class PhyloFlagLoglikelihoodOptions : public PhyloFlagOptions {
+ public:
+  static PhyloFlagLoglikelihoodOptions CreatePhyloFlagLoglikelihoodOptions();
+
+ public:
+  // Exclude options.
+  PhyloFlagName exclude_log_det_jacobian_likelihood_;
+  // Set of All LogLikelihood options.
+  std::set<PhyloFlagName> all_options_;
+};
+
+// ** Map Keys for FatBeagle::Gradient().
+class PhyloFlagGradientMapKeys : public PhyloFlagOptions {
+ public:
+  static PhyloFlagGradientMapKeys CreatePhyloFlagGradientMapKeys();
+
+ public:
+  // Gradient Mapkeys.
+  PhyloFlagName site_model_;
+  PhyloFlagName ratios_root_height_;
+  PhyloFlagName substitution_model_;
+  PhyloFlagName substitution_model_rates_;
+  PhyloFlagName substitution_model_frequencies_;
+  PhyloFlagName site_model_parameters_;
+  PhyloFlagName clock_model_rates_;
+  PhyloFlagName entire_substitution_model_;
+  PhyloFlagName entire_site_model_;
+  PhyloFlagName entire_clock_model_;
+  // Vector of All Gradient map keys.
+  std::set<PhyloFlagName> all_keys_;
+};
+
+// ** All User Options
+class PhyloFlags {
+ public:
+  PhyloFlags(bool is_run_defaults = true);
+  PhyloFlags(const StringVector &key_vec, bool is_run_defaults = true);
+  PhyloFlags(const StringDoubleVector &key_vec, bool is_run_defaults = true);
+
+  // ** Find Flag
+  // Find flag by name.
+  std::optional<PhyloFlagName> FindFlagByName(const std::set<PhyloFlagName> &options,
+                                              const std::string &flag_name,
+                                              std::optional<PhyloFlagName> flag) const;
+  // Find flag by id.
+  std::optional<PhyloFlagName> FindFlagById(const std::set<PhyloFlagName> &options,
+                                            const size_t &flag_id,
+                                            std::optional<PhyloFlagName> flag) const;
+
+  // ** Flag Checked
```

Comment:
At our meeting tomorrow, I'd like to talk a little about this. I'm probably not understanding, but it seems like we could just actually set the values in the default and have a single accessor.

Reply:
Refactor removed "Check" function.